tests/harness/cases/messages.proto

```diff
@@ -28,3 +32,18 @@ message MessageSkip { TestMsg val = 1 [(validate.rules).message.skip = true];}
 message MessageRequired { TestMsg val = 1 [(validate.rules).message.required = true]; }
 
 message MessageWith3dInside {}
+
+message ComplexTestMsg {
+    string                     const  = 1 [(validate.rules).string.const = "abcd"];
+    ComplexTestMsg             nested = 2;
+    int32                      int_const = 3 [(validate.rules).int32.const = 5];
+    bool                       bool_const = 4 [(validate.rules).bool.const = false];
+    google.protobuf.FloatValue float_val = 5 [(validate.rules).float.gt = 0];
+    google.protobuf.Duration   dur_val = 6 [(validate.rules).duration.lt = {seconds: 17}, (validate.rules).duration.required = true];
+    google.protobuf.Timestamp  ts_val = 7 [(validate.rules).timestamp.gt = {seconds: 7}];
+    TestMsg                    another = 8;
+    float                      float_const = 9 [(validate.rules).float.lt = 8];
+    double                     double_in = 10 [(validate.rules).double = {in: [456.789, 123]}];
+}
+
+message ComplexMessage { ComplexTestMsg val = 1; }
```

Comment:
To make it easier to see generated code, this is an example of a message with multiple fields & rules and its generated code:

```
# Validates ComplexMessage
def validate(p):

    if _has_field(p, "val"):
        embedded = generate_validate(p.val)(p.val)
        if embedded is not None:
            return embedded
    return None

# Validates ComplexTestMsg
def validate(p):

    if p.const != "abcd":
        raise ValidationFailed("const not equal to abcd")

    if _has_field(p, "nested"):
        embedded = generate_validate(p.nested)(p.nested)
        if embedded is not None:
            return embedded
    if p.int_const != 5:
        raise ValidationFailed("int_const not equal to 5")

    if p.bool_const != False:
        raise ValidationFailed("bool_const not equal to False")

    if p.HasField("float_val"):
            if p.float_val.value <= 0.0:
                raise ValidationFailed("float_val.value is not greater than 0.0")

    if not _has_field(p, "dur_val"):
        raise ValidationFailed("dur_val is required.")
    if _has_field(p, "dur_val"):
        dur = p.dur_val.seconds + round((10**-9 * p.dur_val.nanos), 9)
        lt = 17.0
        if dur >= lt:
            raise ValidationFailed("dur_val is not lesser than 17.0")

    if _has_field(p, "ts_val"):
        ts = p.ts_val.seconds + round((10**-9 * p.ts_val.nanos), 9)
        gt = 7.0

        if ts <= gt:
            raise ValidationFailed("ts_val is not greater than 7.0")

    if _has_field(p, "another"):
        embedded = generate_validate(p.another)(p.another)
        if embedded is not None:
            return embedded

    if p.float_const >= 8.0:
        raise ValidationFailed("float_const is not lesser than 8.0")

    if p.double_in not in [456.789, 123.0]:
        raise ValidationFailed("double_in not in [456.789, 123.0]")
    return None

# Validates TestMsg
def validate(p):

    if p.const != "foo":
        raise ValidationFailed("const not equal to foo")

    if _has_field(p, "nested"):
        embedded = generate_validate(p.nested)(p.nested)
        if embedded is not None:
            return embedded
    return None
```

Reply:
if all the messages generate a function called `validate` won't that collide?

Reply:
Actually for a message with embedded messages like `message ComplexMessage { ComplexTestMsg val = 1; }`, all that would be returned is:
```
# Validates ComplexMessage
def validate(p):

    if _has_field(p, "val"):
        embedded = generate_validate(p.val)(p.val)
        if embedded is not None:
            return embedded
    return None
```

As seen above, `generate_validate(p.val)(p.val)` is being called on the fields of ComplexMessage `p` and would raise a ValidationFailed exception if it fails to validate.